src/Engine/ProtoCore/DSASM/Heap.cs

```diff
@@ -624,40 +624,45 @@ private bool TryFindFreeIndex(out int index)
             }
         }
 
+        //cache ClassIndex and ProcdureNode for repeated calls of the same type object.
+        private int previousClassIndex;
+        private ProcedureNode pn;
+
         private void GCDisposeObject(StackValue svPtr, Executive exe)
         {
             int classIndex = svPtr.metaData.type;
-            ClassNode cn = exe.exe.classTable.ClassNodes[classIndex];
 
-            ProcedureNode pn = cn.GetDisposeMethod();
-            while (pn == null)
+            if (this.pn == null || classIndex != previousClassIndex)
             {
-                if (cn.Base != Constants.kInvalidIndex)
-                {
-                    classIndex = cn.Base;
-                    cn = exe.exe.classTable.ClassNodes[classIndex];
-                    pn = cn.GetDisposeMethod();
-                }
-                else
+                previousClassIndex = classIndex;
+                ClassNode cn = exe.exe.classTable.ClassNodes[classIndex];
+
+                pn = cn.GetDisposeMethod();
+                while (pn == null)
                 {
-                    break;
+                    if (cn.Base != Constants.kInvalidIndex)
+                    {
+                        classIndex = cn.Base;
+                        cn = exe.exe.classTable.ClassNodes[classIndex];
+                        pn = cn.GetDisposeMethod();
+                    }
+                    else
+                    {
+                        break;
+                    }
                 }
             }
 
             if (pn != null)
             {
-                // TODO Jun/Jiong: Use build pointer utilities 
-                exe.rmem.Push(StackValue.BuildArrayDimension(0));
-                exe.rmem.Push(StackValue.BuildPointer(svPtr.Pointer, svPtr.metaData));
-                exe.rmem.Push(StackValue.BuildInt(1));
-                
                 ++exe.RuntimeCore.FunctionCallDepth;
 
                 // TODO: Need to move IsExplicitCall to DebugProps and come up with a more elegant solution for this
                 // fix for IDE-963 - pratapa
                 bool explicitCall = exe.IsExplicitCall;
                 bool tempFlag = explicitCall;
-                exe.Callr(pn.RuntimeIndex, pn.ID, classIndex, ref explicitCall);
+                //exe.Callr(pn.RuntimeIndex, pn.ID, classIndex, ref explicitCall);
+                exe.CallDipose(pn, svPtr, classIndex, ref explicitCall);
```

Comment:
`explicitCall` is not being touched inside of `CallDispose()`? Can we check this? As far as I remember this had some implications in the case of Dispose calls. Is it no longer relevant?

Reply:
cleaned up